### DISCUSSION TITLE
Code for CD4051BE

### DISCUSSION THREAD

**Discussion by Normanras** [tttapa/Control-Surface#403] [Q&A]
I am attempting a midi controller that needs quite a few faders and will need some multiplexers to bring them all together. I have some CD4051BE's laying around that I've used for other projects. However, when I call the CD74HC4051 in the code, it doesn't work. It is wired up correctly and works with non-control surface lib code. 

I've tried calling the general AnalogMultiplex class and functions, but they also don't seem to work. To start, I was using the code on the main page, substituting certain elements, but now I'm not sure if using AnalogMultiplex requires something else?

Regarding the AnalogMultiplex: the address pins on the CD74HC4051 are S0, S1, S2. On the 4051BE as Channels A, B, C on pins 11, 10, and 9 (respectively, on the 4051BE). The code formatting makes the below look unreadable, so forgive the forgoing of that formatting... 

> AnalogMultiplex mux = {A0,  {3, 4, 5} };
> 
> CCPotentiometer volumePotentiometers[] = {
>   { mux.pin(0), { MIDI_CC::General_Purpose_Controller_1 } },
>   { mux.pin(1), { MIDI_CC::General_Purpose_Controller_2 } },
>   { mux.pin(2), { MIDI_CC::General_Purpose_Controller_3 } },
>   { mux.pin(3), { MIDI_CC::General_Purpose_Controller_4 } },
>   { mux.pin(4), { MIDI_CC::General_Purpose_Controller_5 } },
>   { mux.pin(5), { MIDI_CC::General_Purpose_Controller_6 } },
>   { mux.pin(6), { MIDI_CC::General_Purpose_Controller_7 } },

My plan is to then add Bankable element, which is working currently without the multiplex. I'm assuming I can instantiate the IncrementDecrement class and change the mux array to: 
`{{bank, BankType::CHANGE_ADDRESS}, A9, {MIDI_CC::General_Purpose_Controller_1}},` etc etc

If anyone can help with the mux issue and if I need to just pick up a CD74HC4051, that would be helpful!

**Comment by tttapa** [comment]
Please post your complete code between triple back ticks:

~~~md
```cpp
// your code here
```
~~~
It might be useful to post the working non-control surface code as well.

As far as I can tell, the 4051BE is the same chip as a standard CD4051 multiplexer, the BE simply denotes that it's in a PDIP package. If you think this is not the case, please post a link to the datasheet of the exact chip you're working with.

The code and instructions in the [Getting Started](https://tttapa.github.io/Control-Surface-doc/Doxygen/d5/d7d/md_pages_Getting-Started.html) example should work just fine using your 4051BE. Please double check your wiring. The A, B, C pins should be connected to Arduino pins 3, 4, and 5. The common out/in pin should be connected to Arduino pin A0. The inhibit (enable) pin should be connected to ground. (I'm referring to this schematic: https://www.ti.com/lit/ds/symlink/cd4051b.pdf#page=3.)

If you've tried the exact code in the Getting Started example, double checked your wiring and still can't get it to work, please post a clear picture of your wiring.

Are you sure the multiplexer is the issue? Maybe there's a problem with how you're receiving the MIDI data. To make sure this isn't the case, replace the line `USBMIDI_Interface midi;` with `USBDebugMIDI_Interface midi;` and check the output in the Serial Monitor of the Arduino IDE.

**Reply by Normanras** [reply]
Thank you, @tttapa ! I've maxed out my teensy pins by working some LEDs into working wiring/code to push the project further. I am going to wire up the 4051BE again today or tomorrow and try again with your Debug suggestion. If it still doesn't work I'll take pictures and attach all the code as you've recommended. 

The non-CS code basically was formatted in a generic sense of declaring the address pins and analog pin, using pinMode, digitalWrite, etc, with a for loop under the void loop () { } section. I don't believe I used any libraries. Anyway, that likely isn't very helpful, so let me wire up the mux again and get back to you.

**Reply by Normanras** [reply]
@tttapa Pieter - I rewired my breadboard with the CD4051BE. The link you provided above from TI is correct and what I've been using. I realized that I didn't have every input wired up (I have one empty pin that I've now connected to ground), but even with all the connections accounted for, I still couldn't get it to work. Serial monitor shows inconsistent numbers that do not hold steady - for instance, Data 1 will show 0x00, then when I move the slider it will flicker to 0x013, and then return to 0 even if I don't move the slider. I tried instantiating AnalogMultiplex() but I couldn't get it to work, so I think I'm writing it incorrectly. 

Here's the code:

```
#include <Control_Surface.h>  // Include the library
 
USBDebugMIDI_Interface midi;  // Instantiate a MIDI Interface to use

// Three Banks to change the channel of the pot
Bank<3> bank(7); // 1st number is # of banks, 2nd is # of tracks per bank

// Selector to change banks
IncrementDecrementSelectorLEDs<3> bankSelector = {
    bank,       // Bank to manage
    {0, 1},    //push button pins (inc/dec)
    {6, 9, 10}, //LED Pins
};
 
// Instantiate an analog multiplexer
CD74HC4051 mux = {
  A9,       // Analog input pin
  {2, 3, 4} // Address pins A, B ,C
};

Bankable::CCPotentiometer oscpotentiometers[] = {
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(0), { MIDI_CC::General_Purpose_Controller_1 } },
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(1), { MIDI_CC::General_Purpose_Controller_2 } },
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(2), { MIDI_CC::General_Purpose_Controller_3 } },
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(3), { MIDI_CC::General_Purpose_Controller_4 } },
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(4), { MIDI_CC::General_Purpose_Controller_5 } },
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(5), { MIDI_CC::General_Purpose_Controller_6 } },
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(7), { MIDI_CC::General_Purpose_Controller_7 } },
};

CCPotentiometer MainVolume = {A0, {MIDI_CC::Channel_Volume, CHANNEL_1}};
CCPotentiometer SendFX1 = {A1, {MIDI_CC::Effect_Control_1}};
CCPotentiometer SendFX2 = {A2, {MIDI_CC::Effect_Control_2}};

void setup() {
  Control_Surface.begin();  // Initialize the Control Surface
  delay(10);
}
 
void loop() {
  Control_Surface.loop();  // Update the Control Surface
  delay(10);
}
```

Here's a strange part that I can't debug. With the above code, not only does the multiplexer return strange values, but the three separate pots (that worked when the pots were wired directly to teensy pins), worked flawlessly. So despite being wired and coded the same, they now don't work.  However, the LEDs light up and turn off as I click the buttons, which is normal with the Increment/Decrement function. I've attached some screenshots - here are the mux connections:

Pin1, Channel 4: Slide Pot
Pin2, Channel 6: Slide Pot
Pin3, Com I/O: A9 on Teensy LC
Pin4, Channel 7: Slide Pot
Pin5, Channel 5: Slide Pot
Pin6, INH: GND
Pin7, VEE: GND
Pin8, VSS: GND
Pin9, C: 4 on Teensy LC
Pin10, B: 3 on Teensy LC
Pin11, A: 2 on Teensy LC
Pins 12-15, Channels 0-3: Slide Pots
VDD: 3.3v (I also tried 5v based on the data sheet and [this PJRC discussion](https://forum.pjrc.com/threads/25262-4051-multiplexer-not-working-correctly-with-teensy-3-1-but-works-with-arduino). No change in the serial monitor. 

I'm thinking I might as well pickup some CD74HC4051's just to test them out and confirm it works! 

![CD4051BE_1](https://user-images.githubusercontent.com/44226464/108797099-68c14300-7558-11eb-9848-f11d13763df6.jpg)
![CD4051BE_2](https://user-images.githubusercontent.com/44226464/108797102-6959d980-7558-11eb-97f9-80acaf137d6f.jpg)
<img width="411" alt="CD4051BE Pinout" src="https://user-images.githubusercontent.com/44226464/108797123-74ad0500-7558-11eb-92a5-bc3873aede6e.png">

**Reply by tttapa** [reply]
I've tested your exact code on a Teensy 4.1 with a 4067 multiplexer with the fourth address line wired to ground (I don't have any CD4051BEs at hand). Everything seems to work fine and I'm unable to reproduce the problems you describe.

At this point, it might be a good idea to start the wiring from scratch. Since you're seeing values close to zero, maybe the Vcc connection to the potentiometers is faulty or shorted? This may be totally obvious, but I can't see it in your pictures: how is the red bus on the left hand side of the breadboard connected to the one on the other side?

You could test the multiplexer and your wiring by connecting all address lines (A, B, C) to ground (instead of to the Teensy) so it always selects channel 0 and connecting the wiper of a potentiometer to channel 0 (pin 13 of the mux). Then load the AnalogReadSerial Arduino example and turn the pot. If everything is connected correctly, the value in the serial monitor should vary between 0 and 1023 as you move the potentiometer.

As a side note, the `delay(10)` in your code shouldn't be necessary.

**Reply by Normanras** [reply]
Thanks, Pieter! Super grateful for your input and the library as a whole. I can tell many are grateful too. To answer your direct question: the positive strips are connected to the 3.3v pin - one wire to the right side, and then a second wire from the right side to the left. Same with the GND strips. I know these wires are solid (see below) since I did a bunch of wiring and unwiring and re-wiring with the multiplexer and got consistent results of working and not working.

That's good that the code worked for you without an issue. Before I posted my comment yesterday, I stripped and re-wired the breadboard to try and best check/double-check my setup. I also moved the multiplexer to a different spot on the board, just in case a certain part of the board was bad. I will say, wires don't sit as snugly as I would like on this board and I find myself having to push certain wires back down or moving them when they slip out. But those are mostly limited to the sliding pots, I made sure all multiplex wires and teensy wires were put in snugly. 

I'm wondering if it has something to do with the Teensy LC? I'll have to research. 

And thank you for the delay heads up. I realized it was likely pointless (as it wasn't in any of the examples of yours I referenced) but that PJRC forum post mentioned some strange quirk about 4051BE's that need extra time to fully charge for accurate readings. Apparently someone commented that that worked for them... 

Thanks again, Pieter! If you don't mind one last question before closing this: can the AnalogMultiplex class be called as a generic mux class for all other multiplexers or do you have to instantiate the 4067/4051?

**Reply by tttapa** [reply]
> And thank you for the delay heads up. I realized it was likely pointless (as it wasn't in any of the examples of yours I referenced) but that PJRC forum post mentioned some strange quirk about 4051BE's that need extra time to fully charge for accurate readings. Apparently someone commented that that worked for them...

The delay is required between setting the mux address pins and actually reading the signal pin of the mux. Control Surface already adds this delay on all boards except AVRs (they are too slow to notice the delay of the multiplexer). You can see the code here: 
https://github.com/tttapa/Control-Surface/blob/feee931d44d5f20fb7913b171895316e5819549d/src/AH/Hardware/ExtendedInputOutput/AnalogMultiplex.hpp#L238-L248

The full multiplexer code is in that file, it's pretty standard:
https://github.com/tttapa/Control-Surface/blob/feee931d44d5f20fb7913b171895316e5819549d/src/AH/Hardware/ExtendedInputOutput/AnalogMultiplex.hpp#L214-L221

https://github.com/tttapa/Control-Surface/blob/feee931d44d5f20fb7913b171895316e5819549d/src/AH/Hardware/ExtendedInputOutput/AnalogMultiplex.hpp#L250-L261

One thing that might be different between the Teensy LC and other Teensies is the ADC resolution. Control Surface always sets the ADC resolution to the maximum. I've never actually used a Teensy LC, so this might be incorrect, I simply got the information from the Teensy documentation. You could try setting the resolution to 10 instead of 12 and set `HAS_ANALOG_READ_RESOLUTION` to false:
https://github.com/tttapa/Control-Surface/blob/feee931d44d5f20fb7913b171895316e5819549d/src/AH/Hardware/ADCConfig.hpp#L37-L39

Finally, to test the multiplexer class without any of the MIDI stuff, have a look at the AnalogReadSerial example: https://tttapa.github.io/Control-Surface-doc/Doxygen/dc/d1b/1_8AnalogReadSerial_8ino-example.html

> can the AnalogMultiplex class be called as a generic mux class for all other multiplexers or do you have to instantiate the 4067/4051?

Yes, the AnalogMultiplex should work for almost any kind of multiplexer, it simply sets the address lines, enables the multiplexer if necessary and then reads the input pin, I would think that there's not much that could go wrong there, but apparently, that's not the case.  
The 4067/4051 classes are simply aliases for the AnalogMultiplex class with the correct number of address lines:
https://github.com/tttapa/Control-Surface/blob/feee931d44d5f20fb7913b171895316e5819549d/src/AH/Hardware/ExtendedInputOutput/AnalogMultiplex.hpp#L173-L187

**Reply by Normanras** [reply]
Got it. Thank you. I am going to be afk for a few days but will go through this and see if I can get this to work. Especially if it has to do with the LC's ADC resolution, it would be good to note for other's making their own devices with the LC. I'll circle back.

**Reply by Normanras** [reply]
@tttapa Pieter - just closing the loop here. I picked up a couple 4067 muxes and have had absolutely no issues with your code - it works flawlessly and is easily expandable. It has really helped push the project to the next milestone. 

I'll have to dig into my 4051s as now I'm suspecting there is something wrong with them...

**Reply by tttapa** [reply]
Glad to hear you got it working!

**Comment by alomdj00** [comment]
Hola, uso el cd4051be y esta funcionando correctamente.

Es probable que haya algo mal con teensy, a mi en esp32 tuve que agregar un pequeñobcapacitor para estabilizar la entradas

Los cables-jumpers estan perfectos, no estan rotos?

La fuente de alimentacion esta bien?

Si tienes un arduino a mano prueba con ese haber que tal va el multiplexor,

**Reply by Normanras** [reply]
¡Gracias por su respuesto! Por favor desculpame,  mi español no es tan bueno y hablo mejor que escribo... ja. 

Todos los jumpers si están buenos - nuevo y los corte antes de usarlos. Si la fuente de alimentación es bueno - el código funciona sin problema cuando no uso el multiplexer. Tengo teensy's mas grandes y también un Adafruit Feather M4. Tengo que encontrar en mi caja de cosas electrónicas si tengo un arduino. 

Finalmente, donde has colocado los capaticores? en fronte de los potentiometers o cerca de los pins del multiplexer? (no se esta palabra en español).
